### ISSUE TITLE
[PERFORMANCE] Accerlerating KV-Cache during Inference

### ISSUE THREAD

**Issue by yiyousong** [SmallDoges/flash-dmattn#141]
First, congratulations on your outstanding work. However, I checked the paper and code and noticed some minor improvements (or I failed to notice my mistakes). From the code, I believe the scores are always static, and hence the KV-cache can be reduced to linear, because **the evicted KV will never be used again**.

Below is a simple proof.

Denote Score (before topK) as $S=f(V)$
Denote Selected Index (Mask) as $M_N=TopK(S_{1:N}) $
Obviously, $M_N=TopK(S_{1:N})=TopK(TopK([S_{1:N-1}),S_{N}])=TopK(M_{N-1},S_N)$
Which means that for each input, at most one index can be evicted, and no evicted tokens will be selected.

Therefore, during inference, you only need to maintain a KV cache of size window_size.

**Comment by LoserCheems** [comment]
Thank you, @yiyouson. I think in the inference process, for each query, the $K$ most relevant kv caches should be selected.
If some kv caches are removed from the cache list in the previous time step, they will never be included in the selection list for the current time step.

**Comment by LoserCheems** [comment]
I would like to confirm whether, for each query, the most relevant KV cache should be different

**Comment by LoserCheems** [comment]
Thank you @yiyousong for the insightful observation! Your TopK maintenance argument is correct under the assumption that the routing scores (attention_bias in our current integration example) are static with respect to time and query content. In our current code path, the active set is indeed derived from attention_bias before the QK dot-product is computed, so if attention_bias does not change once a token is introduced, an evicted position will never re-enter the TopK set. In that static-bias regime, we can maintain only a bounded KV cache of size keep_window_size (plus the most recent token), updating it incrementally by merging the new token’s score with the previous TopK.

**Comment by LoserCheems** [comment]
However, if we (or downstream users) supply a query-dependent or adaptively updated bias (e.g. a function of the current query, gradient statistics, or other dynamic routing signals), the monotonic eviction property no longer holds and previously evicted tokens might become relevant again. For that dynamic setting we still need the full KV growth or a more elaborate retrieval/indexing scheme.

Planned follow-ups:
1. Add documentation clarifying “static bias TopK” vs “dynamic query-aware TopK”.
2. Provide an optional streaming TopK cache mode that keeps only keep_window_size KV slots and updates them incrementally (O(K) per step).
3. Supply correctness tests showing identical outputs (within tolerance) compared to the full cache under static bias.
4. Benchmark memory & latency improvements on long-context scenarios.

If you (or others) are interested, we can prototype a minimal StreamingTopKCache in a follow-up PR.

Happy to discuss further or review a contribution if you’d like to help implement this mode.

**Comment by yiyousong** [comment]
Actually, I plan to do a small follow-up work like MHA2MLA, where I plan to investigate whether DMA can be finetuned from modern LLMs and also test whether sliding window attention will be useful in this case. Because 
1. the evicted token will not be used again
2. tokens need to focus on recent tokens
which means to maintain any locality in KV-cache, the score needs to be continual increasing. This issue can be solved using sliding window+ score-based cache.

**Comment by yiyousong** [comment]
FYI, default dynamic scores means O(N^2) complexity. You'll have to first group the Vs into groups to reduce the complexity to O(NG). Feels kinda like **cluster attention** based on V instead of K. 
I believe I've seen a somewhat recent paper related to "TopK cluster attention" but I cannot find it, maybe it will be helpful to you

**Comment by yiyousong** [comment]
> I would like to confirm whether, for each query, the most relevant KV cache should be different

Query-based mask could be helpful, but the SM(QK) already handles that, so it won't be THAT important. Plus, your results already showed that it can work. I'd say its only worth testing if you strongly need a tiny cache and got the computation powers for another round of pretraining

**Comment by LoserCheems** [comment]
> Actually, I plan to do a small follow-up work like MHA2MLA, where I plan to investigate whether DMA can be finetuned from modern LLMs and also test whether sliding window attention will be useful in this case. Because
> 
> 1. the evicted token will not be used again
> 2. tokens need to focus on recent tokens
>    which means to maintain any locality in KV-cache, the score needs to be continual increasing. This issue can be solved using sliding window+ score-based cache.

It's truly amazing!

In fact, we discovered in medium-scale language modeling that during the early stage of pre-training, using full attention, and then in the extended context stage, adjusting the RoPE base and enabling DMA while keeping the window size consistent with the sequence length in the early stage, can achieve perfect results.

**Comment by LoserCheems** [comment]
> > I would like to confirm whether, for each query, the most relevant KV cache should be different
> 
> Query-based mask could be helpful, but the SM(QK) already handles that, so it won't be THAT important. Plus, your results already showed that it can work. I'd say its only worth testing if you strongly need a tiny cache and got the computation powers for another round of pretraining

In fact, we are also exploring a strategy to dynamically adjust the window size based on the current query. This way, the DMA can obtain the relevant window for the query and also be able to adjust the content to be calculated according to the values.

**Comment by yiyousong** [comment]
I checked, and DMA (with or without sliding window) cannot be directly converted from a finetuned attention model. I'm gonna stop working on this. 
below is the code to add a sliding window,seems like the current kernel can be re-used without modification
```
attn_mask = attn_mask.scatter(-1, topk_indices, valid_topk.to(dtype))# add windowed mask after this line
attn_mask = attn_mask.masked_fill(
            (torch.arange(attn_bias.shape[-1], device=attn_bias.device) > cache_position.reshape(-1, 1)-sliding_window_size)\
                * (torch.arange(attn_bias.shape[-1], device=attn_bias.device) < cache_position.reshape(-1, 1)+1),
            1
        )
```

**Comment by LoserCheems** [comment]
The most troublesome thing is that our checkpoint was accidentally deleted by a classmate in the group. We can only retrain a new one and re-release it.

**Comment by LoserCheems** [comment]
I plan to pre-train a DMA model of around 200M in the next two weeks, and release the intermediate checkpoints.

**Comment by LoserCheems** [comment]
Hi @yiyousong, please refer to @huangyuxiang03 latest work: https://arxiv.org/pdf/2510.13602
This might solve your problem
